kernel/src/common/regs/README.md

```diff
@@ -0,0 +1,266 @@
+is a field in a packed struct:
+
+```rust
+use common::regs::{ReadOnly, ReadWrite, WriteOnly};
+
+#[repr(C, packed)]
+struct Registers {
+    // Control register: read-write
+    // The 'Control' parameter constrains this register to only use fields from
+    // a certain group (defined below in the bitfields section).
+    cr: ReadWrite<u8, Control>,
+
+    // Status register: read-only
+    s: ReadOnly<u8, Status>
+
+    // Registers can be bytes, halfwords, or words:
+    // Note that the second type parameter can be omitted, meaning that there
+    // are no bitfields defined for these registers.
+    byte0: ReadWrite<u8>,
+    byte1: ReadWrite<u8>,
+    short: ReadWrite<u16>,
+    word: ReadWrite<u32>
+
+    // Etc.
+}
+```
+
+## Defining bitfields
+
+Bitfields are defined through the `bitfields!` macro:
+
+```rust
+bitfields! [
+    // First parameter is the register width for the bitfields. Can be u8, u16,
+    // or u32.
+    u8,
+
+    // Each subsequent parameter is a register abbreviation, its descriptive
+    // name, and its associated bitfields.
+    // The descriptive name defines this 'group' of bitfields. Only registers
+    // defined as ReadWrite<_, Control> can use these bitfields.
+    CR Control [
+        // Bitfields are defined as:
+        // name (Mask(mask), shift) [ /* optional named values */ ]
+
+        // This is a two-bit field which includes bits 4 and 5
+        RANGE (Mask(0b11), 4) [
+            // Each of these defines a name for a value that the bitfield can be
+            // written with or matched against. Note that this set is not exclusive--
+            // the field can still be written with arbitrary constants.
+            VeryHigh = 0,
+            High = 1,
+            Low = 2
+        ],
+
+        // A common case is single-bit bitfields, which usually just mean
+        // 'enable' or 'disable' something. This syntax is a shorthand, defining
+        // EN as a field over bit 3, and INT as a field over bit 2.
+        EN 3 [],
+        INT 2 []
+    ],
+
+    // Without the explanatory comments like above, bitfield definition is quite compact:
+    // Status register
+    S Status [
+        TXCOMPLETE 0 [],
+        TXINTERRUPT 1 [],
+        RXCOMPLETE 2 [],
+        RXINTERRUPT 3 [],
+        MODE (Mask(0b11), 4) [
+            FullDuplex = 0,
+            HalfDuplex = 1,
+            Loopback = 2,
+            Disabled = 3
+        ],
+        ERRORCOUNT (Mask(0b11), 6) []
+    ]
+]
+```
+
+## Register Interface Summary
+
+There are three types provided by the register interface: `ReadOnly`, `WriteOnly`, and `ReadWrite`. They provide the following functions:
+
+```rust
+ReadOnly<T: IntLike, R: RegisterLongName = ()>
+.get() -> T                                 // Get the raw register value
+.read(field: Field<T, R>) -> T              // Read the value of the given field
+.is_set(field: Field<T, R>) -> bool         // Check if a one-bit field is set
```

Comment:
What happens if you call `is_set` on a two (or more) bit field? Does it fail to compile?

Reply:
Right now, it would just return true if the field is nonzero, and false otherwise. However, because there's a shorthand for defining 1-bit fields, it would be possible to specialize the is_set for a 1-bit specific version of ReadOnly/ReadWrite

Reply:
Returning true if any bit is one seems reasonable. It is "set" to some value.

Reply:
Could rename to `is_nonzero` and add a complementary `is_zero` method instead of `is_set`

Reply:
Don't feel strongly about renaming or not, but if we leave `is_set`, I think the comment `// Check if a one-bit field is set` should change to `// Check if any bits are set` ?

Reply:
Updated comment.